File: apps/android/app/src/test/java/com/reverie/app/domain/search/QueryTokensTest.kt
package com.reverie.app.domain.search

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Test

/**
 * Ported verbatim from libs/shared/src/search/query-tokens.test.ts. These fixtures are the
 * contract: the Kotlin DSL must produce identical `q` strings to the web/backend so the
 * server re-parses them the same way.
 */
class QueryTokensTest {

    private fun text(value: String, negated: Boolean = false, raw: String = value) =
        QueryToken(TokenType.TEXT, value, negated = negated, raw = raw)

    private fun quoted(value: String, negated: Boolean = false, raw: String) =
        QueryToken(TokenType.QUOTED, value, negated = negated, raw = raw)

    private fun filter(key: String, value: String, negated: Boolean = false, raw: String) =
        QueryToken(TokenType.FILTER, value, key = key, negated = negated, raw = raw)

    // ---- tokenizeQuery ----

    @Test fun `tokenizes free text words`() {
        assertEquals(
            listOf(text("beach"), text("sunset")),
            tokenizeQuery("beach sunset"),
        )
    }

    @Test fun `tokenizes quoted free text`() {
        assertEquals(
            listOf(quoted("beach sunset", raw = "\"beach sunset\"")),
            tokenizeQuery("\"beach sunset\""),
        )
    }

    @Test fun `tokenizes filters and lowercases keys`() {
        assertEquals(
            listOf(filter("type", "photo", raw = "TYPE:photo")),
            tokenizeQuery("TYPE:photo"),
        )
    }

    @Test fun `tokenizes negated filters`() {
        assertEquals(
            listOf(filter("has", "text", negated = true, raw = "-has:text")),
            tokenizeQuery("-has:text"),
        )
    }

    @Test fun `handles quoted filter values with whitespace`() {
        assertEquals(
            listOf(
                filter("entity", "John Smith", raw = "entity:\"John Smith\""),
                text("beach"),
            ),
            tokenizeQuery("entity:\"John Smith\" beach"),
        )
    }

    @Test fun `handles single-word quoted filter values`() {
        assertEquals(
            listOf(filter("folder", "Seppo", raw = "folder:\"Seppo\"")),
            tokenizeQuery("folder:\"Seppo\""),
        )
    }

    @Test fun `consumes unterminated quotes to end of string`() {
        assertEquals(
            listOf(quoted("beach sun", raw = "\"beach sun")),
            tokenizeQuery("\"beach sun"),
        )
        assertEquals(
            listOf(filter("entity", "John Smi", raw = "entity:\"John Smi")),
            tokenizeQuery("entity:\"John Smi"),
        )
    }

    @Test fun `keeps empty filter values`() {
        assertEquals(
            listOf(filter("tag", "", raw = "tag:")),
            tokenizeQuery("tag:"),
        )
    }

    @Test fun `treats a leading colon as text`() {
        assertEquals(
            listOf(text(":value")),
            tokenizeQuery(":value"),
        )
    }

    @Test fun `preserves value case`() {
        assertEquals("Important", tokenizeQuery("tag:Important")[0].value)
    }

    @Test fun `handles date and size values as opaque strings`() {
        assertEquals(
            listOf(
                filter("uploaded", "2024-01..2024-06", raw = "uploaded:2024-01..2024-06"),
                filter("size", ">10MB", raw = "size:>10MB"),
            ),
            tokenizeQuery("uploaded:2024-01..2024-06 size:>10MB"),
        )
    }

    // ---- serializeQuery ----

    @Test fun `round-trips exactly`() {
        val queries = listOf(
            "beach type:photo -has:text",
            "entity:\"John Smith\" tag:tax uploaded:last-week",
            "\"exact phrase\" folder:/vacation/2024",
            "type:photo type:video size:>10MB",
        )
        for (q in queries) {
            assertEquals(q, serializeQuery(tokenizeQuery(q)))
        }
    }

    @Test fun `is stable on a second pass`() {
        val q = "beach   type:photo    entity:\"John Smith\""
        val once = serializeQuery(tokenizeQuery(q))
        assertEquals(once, serializeQuery(tokenizeQuery(once)))
    }

    // ---- isKnownFilter / getFreeText / getFilterTokens ----

    @Test fun `marks unknown keys as free text`() {
        val tokens = tokenizeQuery("beach foo:bar type:photo")
        assertFalse(isKnownFilter(tokens[1]))
        assertEquals("beach foo:bar", getFreeText(tokens))
        assertEquals(listOf("type"), getFilterTokens(tokens).map { it.key })
    }

    @Test fun `excludes bare star from free text`() {
        assertEquals("", getFreeText(tokenizeQuery("* type:photo")))
    }

    @Test fun `filters tokens by key`() {
        val tokens = tokenizeQuery("tag:a tag:b type:photo")
        assertEquals(listOf("a", "b"), getFilterTokens(tokens, FilterKey.TAG).map { it.value })
    }

    // ---- addFilter ----

    @Test fun `appends a filter token`() {
        assertEquals("beach type:photo", addFilter("beach", FilterKey.TYPE, "photo"))
    }

    @Test fun `quotes values with whitespace`() {
        assertEquals("entity:\"John Smith\"", addFilter("", FilterKey.ENTITY, "John Smith"))
    }

    @Test fun `dedupes identical tokens`() {
        assertEquals("type:photo", addFilter("type:photo", FilterKey.TYPE, "photo"))
    }

    @Test fun `adds negated tokens distinctly`() {
        assertEquals("has:text -has:text", addFilter("has:text", FilterKey.HAS, "text", negated = true))
    }

    // ---- removeFilter ----

    @Test fun `removes by key and value at token level`() {
        assertEquals(
            "entity:\"tax office\"",
            removeFilter("tag:tax entity:\"tax office\"", FilterKey.TAG, "tax"),
        )
    }

    @Test fun `does not substring-match across keys`() {
        assertEquals("format:pdf", removeFilter("tag:pdf format:pdf", FilterKey.TAG, "pdf"))
    }

    @Test fun `removes all tokens of a key when value omitted`() {
        assertEquals("beach", removeFilter("tag:a tag:b beach", FilterKey.TAG))
    }

    @Test fun `matches values case-insensitively`() {
        assertEquals("", removeFilter("folder:Seppo", FilterKey.FOLDER, "seppo"))
    }

    // ---- replaceFilter ----

    @Test fun `replaces all tokens of a key with one value`() {
        assertEquals(
            "beach uploaded:last-week",
            replaceFilter("uploaded:2024 beach", FilterKey.UPLOADED, "last-week"),
        )
    }

    // ---- setFreeText ----

    @Test fun `replaces free text and preserves filters`() {
        assertEquals("sunset type:photo", setFreeText("beach type:photo", "sunset"))
    }

    @Test fun `clears free text when empty`() {
        assertEquals("type:photo", setFreeText("beach type:photo", ""))
    }

    @Test fun `lifts filter syntax typed into the text`() {
        assertEquals("beach tag:summer type:photo", setFreeText("type:photo", "beach tag:summer"))
    }

    @Test fun `drops unknown-key pseudo filters with the old text`() {
        assertEquals("beach type:photo", setFreeText("foo:bar type:photo", "beach"))
    }
}

File: libs/shared/src/search/query-tokens.test.ts
import { describe, expect, it } from 'vitest';
import {
    addFilter,
    getFilterTokens,
    getFreeText,
    isKnownFilter,
    removeFilter,
    replaceFilter,
    serializeQuery,
    setFreeText,
    tokenizeQuery,
} from './query-tokens.js';

describe('tokenizeQuery', () => {
    it('tokenizes free text words', () => {
        expect(tokenizeQuery('beach sunset')).toEqual([
            { type: 'text', value: 'beach', negated: false, raw: 'beach' },
            { type: 'text', value: 'sunset', negated: false, raw: 'sunset' },
        ]);
    });

    it('tokenizes quoted free text', () => {
        expect(tokenizeQuery('"beach sunset"')).toEqual([{ type: 'quoted', value: 'beach sunset', negated: false, raw: '"beach sunset"' }]);
    });

    it('tokenizes filters and lowercases keys', () => {
        expect(tokenizeQuery('TYPE:photo')).toEqual([{ type: 'filter', key: 'type', value: 'photo', negated: false, raw: 'TYPE:photo' }]);
    });

    it('tokenizes negated filters', () => {
        expect(tokenizeQuery('-has:text')).toEqual([{ type: 'filter', key: 'has', value: 'text', negated: true, raw: '-has:text' }]);
    });

    it('handles quoted filter values with whitespace', () => {
        expect(tokenizeQuery('entity:"John Smith" beach')).toEqual([
            { type: 'filter', key: 'entity', value: 'John Smith', negated: false, raw: 'entity:"John Smith"' },
            { type: 'text', value: 'beach', negated: false, raw: 'beach' },
        ]);
    });

    it('handles single-word quoted filter values', () => {
        expect(tokenizeQuery('folder:"Seppo"')).toEqual([{ type: 'filter', key: 'folder', value: 'Seppo', negated: false, raw: 'folder:"Seppo"' }]);
    });

    it('consumes unterminated quotes to end of string', () => {
        expect(tokenizeQuery('"beach sun')).toEqual([{ type: 'quoted', value: 'beach sun', negated: false, raw: '"beach sun' }]);
        expect(tokenizeQuery('entity:"John Smi')).toEqual([{ type: 'filter', key: 'entity', value: 'John Smi', negated: false, raw: 'entity:"John Smi' }]);
    });

    it('keeps empty filter values', () => {
        expect(tokenizeQuery('tag:')).toEqual([{ type: 'filter', key: 'tag', value: '', negated: false, raw: 'tag:' }]);
    });

    it('treats a leading colon as text', () => {
        expect(tokenizeQuery(':value')).toEqual([{ type: 'text', value: ':value', negated: false, raw: ':value' }]);
    });

    it('preserves value case (semantic lowercasing is backend-side)', () => {
        expect(tokenizeQuery('tag:Important')[0]?.value).toBe('Important');
    });

    it('handles date and size values as opaque strings', () => {
        expect(tokenizeQuery('uploaded:2024-01..2024-06 size:>10MB')).toEqual([
            { type: 'filter', key: 'uploaded', value: '2024-01..2024-06', negated: false, raw: 'uploaded:2024-01..2024-06' },
            { type: 'filter', key: 'size', value: '>10MB', negated: false, raw: 'size:>10MB' },
        ]);
    });
});

describe('serializeQuery', () => {
    it('round-trips exactly (raw preservation)', () => {
        const queries = [
            'beach type:photo -has:text',
            'entity:"John Smith" tag:tax uploaded:last-week',
            '"exact phrase" folder:/vacation/2024',
            'type:photo type:video size:>10MB',
        ];

        for (const q of queries) {
            expect(serializeQuery(tokenizeQuery(q))).toBe(q);
        }
    });

    it('is stable on a second pass', () => {
        const q = 'beach   type:photo    entity:"John Smith"';
        const once = serializeQuery(tokenizeQuery(q));

        expect(serializeQuery(tokenizeQuery(once))).toBe(once);
    });
});

describe('isKnownFilter / getFreeText / getFilterTokens', () => {
    it('marks unknown keys as free text', () => {
        const tokens = tokenizeQuery('beach foo:bar type:photo');

        expect(isKnownFilter(tokens[1]!)).toBe(false);
        expect(getFreeText(tokens)).toBe('beach foo:bar');
        expect(getFilterTokens(tokens).map((t) => t.key)).toEqual(['type']);
    });

    it('excludes bare * from free text', () => {
        expect(getFreeText(tokenizeQuery('* type:photo'))).toBe('');
    });

    it('filters tokens by key', () => {
        const tokens = tokenizeQuery('tag:a tag:b type:photo');

        expect(getFilterTokens(tokens, 'tag').map((t) => t.value)).toEqual(['a', 'b']);
    });

    it('recognizes the content filter key (not free text)', () => {
        const tokens = tokenizeQuery('content:"invoice number" beach');

        expect(isKnownFilter(tokens[0]!)).toBe(true);
        expect(getFreeText(tokens)).toBe('beach');
        expect(getFilterTokens(tokens, 'content').map((t) => t.value)).toEqual(['invoice number']);
    });
});

describe('addFilter', () => {
    it('appends a filter token', () => {
        expect(addFilter('beach', 'type', 'photo')).toBe('beach type:photo');
    });

    it('quotes values with whitespace', () => {
        expect(addFilter('', 'entity', 'John Smith')).toBe('entity:"John Smith"');
    });

    it('dedupes identical tokens', () => {
        expect(addFilter('type:photo', 'type', 'photo')).toBe('type:photo');
    });

    it('adds negated tokens distinctly', () => {
        expect(addFilter('has:text', 'has', 'text', { negated: true })).toBe('has:text -has:text');
    });
});

describe('removeFilter', () => {
    it('removes by key and value at token level', () => {
        expect(removeFilter('tag:tax entity:"tax office"', 'tag', 'tax')).toBe('entity:"tax office"');
    });

    it('does not substring-match across keys (format:pdf vs tag:pdf)', () => {
        expect(removeFilter('tag:pdf format:pdf', 'tag', 'pdf')).toBe('format:pdf');
    });

    it('removes all tokens of a key when value omitted', () => {
        expect(removeFilter('tag:a tag:b beach', 'tag')).toBe('beach');
    });

    it('matches values case-insensitively', () => {
        expect(removeFilter('folder:Seppo', 'folder', 'seppo')).toBe('');
    });
});

describe('replaceFilter', () => {
    it('replaces all tokens of a key with one value', () => {
        expect(replaceFilter('uploaded:2024 beach', 'uploaded', 'last-week')).toBe('beach uploaded:last-week');
    });
});

describe('setFreeText', () => {
    it('replaces free text and preserves filters', () => {
        expect(setFreeText('beach type:photo', 'sunset')).toBe('sunset type:photo');
    });

    it('clears free text when empty', () => {
        expect(setFreeText('beach type:photo', '')).toBe('type:photo');
    });

    it('lifts filter syntax typed into the text', () => {
        expect(setFreeText('type:photo', 'beach tag:summer')).toBe('beach tag:summer type:photo');
    });

    it('drops unknown-key pseudo filters with the old text', () => {
        expect(setFreeText('foo:bar type:photo', 'beach')).toBe('beach type:photo');
    });
});
